packages/core/contracts/oracle/implementation/VotingV2.sol

```diff
@@ -923,27 +938,38 @@ contract VotingV2 is
                 slash += int256(((voterStake.activeStake * totalSlashed)) / totalCorrectVotes);
             }
 
-            // If this is not the last price request to apply and the next request in the batch is from a subsequent
-            // round then apply the slashing now. Else, do nothing and apply the slashing after the loop concludes.
-            // This acts to apply slashing within a round as independent actions: multiple votes within the same round
-            // should not impact each other but subsequent rounds should impact each other. We need to consider the
-            // skippedRequestIndexes mapping when finding the next index as the next request may have been deleted or rolled.
-            uint256 nextRequestIndex =
-                skippedRequestIndexes[requestIndex + 1] != 0
-                    ? skippedRequestIndexes[requestIndex + 1] + 1
-                    : requestIndex + 1;
+            nextIndexToProcess = requestIndex + 1;
+        }
+
+        // If there is any remaining slashing then apply it. This occurs when there is unapplied slashing in the loop
+        // due to the last unlashed elements all being all from the same round. i.e we only slash within the loop when
+        // transitioning between rounds and the last round is slashed here. Note that there is a special case that needs
+        // to be considered separately: if the nextIndex that we're going to process is >= priceRequestIds, then we
+        // know that there's going to be a round change because new requests never get added to a past round.
+        // If we are not in this case and the next element to be processed has the same round, then we know that
+        // we've bisected a round and should store the unapplied slashing which will seed this method on the next entry
+        // such that the slashing will be applied linearly, not compounding with other slashing within the same round.
+        if (slash != 0) {
+            // The next index could be either a the result of the skip for the next value if it's nonzero or just the
+            // next unprocessed index if there is no skip value for it. This ensures that the price request we read has
+            // not been modified by round-changing when rolling.
+            uint256 nextIndex =
+                skippedRequestIndexes[nextIndexToProcess] != 0
+                    ? skippedRequestIndexes[nextIndexToProcess] + 1
+                    : nextIndexToProcess;
```

Comment:
This is all explained in the comment, but the next index can either be a jump or not.

If it's a jump, it _must_ always land on an element that hasn't had its round modified. If it's skipped bc of spam, it's round shouldn't be modified, so that will correctly detect a round change*.

If it's _not a jump_, we know the next request has not had its round modified, so we can use it directly.

*There is a tiny edge case that I don't think matters, but should be mentioned. If you have a roll skip that is adjacent to a spam skip (or adjacent spam skips for whatever reason), you can have this code _not think_ there's going to be a round change when there really is going to be a round change. If the roll skip lands on something in the same round, but then the spam skip skips to the following round, this code will think there's no round change, because the beginning of the spam skip is in the same round. However, the next element to be processed will ultimately be a different round. The reason I think this doesn't matter is because the only effect of this will be that the slashing will still be applied correctly once this loop catches up with the end of the array. However, we will end up not detecting that we're about to change rounds and only recognize that on the next run. Net result: we store unapplied slash in this rare case despite it being technically possible to go ahead and apply that slash now and avoid storing that extra variable. Please let me know if you think this matters. I don't think it should affect anything other than increase the gas costs in this exceedingly rare case.

Reply:
thank you for adding the detailed explanation. this is subtle.

regarding your edge case I think you right. it's just an extra no-op. would be curious to here if OZ agrees with this. I'll link to this comment in particular in the review doc.